src/blob_file_set.cc

```diff
@@ -209,25 +210,100 @@ Status BlobFileSet::WriteSnapshot(log::Writer* log) {
   return s;
 }
 
+// A helper class to collect all the information needed for manifest write.
+struct BlobFileSet::ManifestWriter {
+  Status status;
+  bool done;
+  port::CondVar cv;
+  EditCollector& collector;
+  VersionEdit& edit;
+
+  explicit ManifestWriter(port::Mutex* mu, EditCollector& _collector,
+                          VersionEdit& _edit)
+      : done(false), cv(mu), collector(_collector), edit(_edit) {}
+};
+
 Status BlobFileSet::LogAndApply(VersionEdit& edit) {
+  mutex_->AssertHeld();
   TEST_SYNC_POINT("BlobFileSet::LogAndApply");
-  // TODO(@huachao): write manifest file unlocked
-  std::string record;
   edit.SetNextFileNumber(next_file_number_.load());
-  edit.EncodeTo(&record);
 
   EditCollector collector;
   Status s = collector.AddEdit(edit);
   if (!s.ok()) return s;
   s = collector.Seal(*this);
   if (!s.ok()) return s;
-  s = manifest_->AddRecord(record);
-  if (!s.ok()) return s;
 
-  ImmutableDBOptions ioptions(db_options_);
-  s = SyncTitanManifest(stats_, &ioptions, manifest_->file());
-  if (!s.ok()) return s;
-  return collector.Apply(*this);
+  ManifestWriter writer(mutex_, collector, edit);
+  manifest_writers_.push_back(&writer);
+
+  // The head of queue is the writer that is responsible for writing manifest
+  // for the group. Thw write group makes sure only one thread is writing to the
+  // manifest, otherwise the manifest will be corrupted.
+  while (!writer.done && &writer != manifest_writers_.front()) {
+    writer.cv.Wait();
+  }
+  if (writer.done) {
+    // The writer is handled by the head of queue's writer, just return here
+    return writer.status;
```

Comment:
manifest_writers_ will hold a reference of writer that is deleted  after the function exists.

Reply:
The writers must be popped up from `manifest_writers_` by the writer ahead of queue when it reaches here.